src/cbmc_starter_kit/ctagst.py

```diff
@@ -0,0 +1,158 @@
+        logging.info("Running universal ctags")
+        stdout, _ = popen(cmd, cwd=root, stdin='\n'.join(files))
+        strings = stdout.splitlines()
+    except UserWarning:
+        logging.info("Universal ctags failed")
+        strings = []
+
+    return [tag for string in strings for tag in universal_tag(root, string)]
+
+def universal_tag(root, string):
+    """Extract tag from universal ctag output."""
+
+    try:
+        # universal ctag json output is '{"name": symbol, "path": file, "line": line, "kind": kind}'
+        blob = json.loads(string)
+        return [{'symbol': blob['name'], 'file': root/blob['path'], 'line': int(blob['line']),
+                 'kind': blob['kind']}]
+    except (json.decoder.JSONDecodeError, # json is unparsable
+            KeyError,                     # json key is missing
+            ValueError) as error:         # invalid literal for int()
+        logging.debug("Bad universal ctag: %s: %s", string, error)
+        return []
+
+################################################################
+
+def exhuberant_ctags(root, files):
+    """Use exhuberant ctags to list symbols defined in files under root."""
+
+    # See exhuberant ctags man page at https://linux.die.net/man/1/ctags
+    cmd = [
+        'ctags',
+        '-L', '-', # read files from standard input, one file per line
+        '-f', '-', # write tags to standard output, one tag per line
+        '-n', # use line numbers (not search expressions) to locate symbol in file
+        '--fields=K' # include symbol kind among extension fields
+    ]
+    try:
+        logging.info("Running exhuberant ctags")
+        stdout, _ = popen(cmd, cwd=root, stdin='\n'.join(files))
+        strings = stdout.splitlines()
+    except UserWarning:
+        logging.info("Exhuberant ctags failed")
+        strings = []
+
+    return [tag for string in strings for tag in exhuberant_tag(root, string)]
+
+def exhuberant_tag(root, string):
+    """Extract tag from exhuberant ctag output."""
+
+    try:
+        # exhuberant ctag output is 'symbol<TAB>path<TAB>line;"<TAB>kind'
+        left, right = string.split(';"')[:2]
+        symbol, path, line = left.split("\t")[:3]
+        kind = right.split("\t")[1]
+        return [{'symbol': symbol, 'file': root/path, 'line': int(line), 'kind': kind}]
+    except (ValueError, IndexError): # not enough values to unpack, invalid literal for int()
+        logging.debug('Bad exhuberant ctag: "%s"', string)
+        return []
+
+################################################################
+
+def legacy_ctags(root, files):
+    """Use legacy ctags to list symbols defined in files under root."""
+
+    # MacOS ships with a legacy ctags from BSD installed in /usr/bin/ctags.
+    # See the MacOS man page for the documentation used to implement this method.
+    cmd = ['ctags',
+           '-x',  # write human-readable summary to standard output
+           *files # legacy ctags cannot read list of files from stdin
```

Comment:
Is the list of files ever expected to be large? The operating system has a maximum length for argument lists, and it varies between OSs. I guess the catch will deal with this failing to launch though.
```
# on macOS
; getconf ARG_MAX
1048576
```

Reply:
The invocation will fail, and the only symbols that will be linked are the symbols that actually appear in the cbmc symbol table.  That's not so awful.  And the right solution is to install a more modern version of ctags (the legacy is from BSD 3.0 and earlier).